src/kaleidoscope/device/olkb/Planck.h

```diff
@@ -20,33 +20,29 @@
 #ifdef ARDUINO_AVR_PLANCK
 
 #include <Arduino.h>
-#define HARDWARE_IMPLEMENTATION kaleidoscope::hardware::olkb::Planck
-#include "Kaleidoscope-HIDAdaptor-KeyboardioHID.h"
+#define DEVICE_CLASS_NAME kaleidoscope::device::olkb::Planck
 
-#include "kaleidoscope/macro_helpers.h"
-#include "kaleidoscope/hardware/avr/pins_and_ports.h"
-
-#include "kaleidoscope/hardware/ATMegaKeyboard.h"
+#include "kaleidoscope/driver/keyscanner/AVR.h"
+#include "kaleidoscope/driver/bootloader/avr/HalfKay.h"
+#include "kaleidoscope/device/avr/ATMega32U4.h"
 
 namespace kaleidoscope {
-namespace hardware {
+namespace device {
 namespace olkb {
-class Planck: public kaleidoscope::hardware::ATMegaKeyboard {
- public:
-  Planck(void) {}
-
-  ATMEGA_KEYBOARD_CONFIG(
-    ROW_PIN_LIST({ PIN_D0, PIN_D5, PIN_B5, PIN_B6 }),
-    COL_PIN_LIST({ PIN_F1, PIN_F0, PIN_B0, PIN_C7, PIN_F4, PIN_F5, PIN_F6, PIN_F7, PIN_D4, PIN_D6, PIN_B4, PIN_D7 })
-  );
-
-  static constexpr int8_t led_count = 0;
 
-  static constexpr int8_t numKeys() {
-    return matrix_columns * matrix_rows;
-  }
+struct PlanckProps : kaleidoscope::device::avr::ATMega32U4Props {
+  struct KeyScannerProps : public kaleidoscope::driver::keyscanner::AVRProps {
```

Comment:
I'm a little worried about going too far with the macro. The situation I hope to avoid is having to replace the macro with code when you need to do something a little bit more complicated.

I'm ok with swapping the macro for "full" code when you go from a slightly complicated device to a very complicated device, but I don't want there to be a big cliff for folks who need something only a bit custom.

Reply:
I've created a macro for now. It doesn't support "only a bit of custom", but there are examples for both macro and macro-less boards (both very simple ones), the cliff shouldn't be big at all. It's rather a tiny bump in the road.